### PR TITLE
feat(chartViewport): segmentation :: 4D chart viewport

### PR DESCRIPTION
**Pull request by lscoder**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -88,6 +88,7 @@
     "copy-webpack-plugin": "^9.0.1",
     "cross-env": "^5.2.0",
     "css-loader": "^3.2.0",
+    "d3": "^7.8.5",
```
Comment (sedghi) [inline package.json:91]: do we really need all of d3?

File: platform/core/src/services/WorkflowStepsService/WorkflowStepsService.ts [hunk 2]
```diff
@@ -165,23 +169,39 @@ class WorkflowStepsService extends PubSubService {
...
-    this._updatePanels(activeWorkflowStep);
-    this._updateHangingProtocol(activeWorkflowStep);
-    this._broadcastEvent(EVENTS.ACTIVE_STEP_CHANGED, { activeWorkflowStep });
+    const appContext = {
+      extensionManager: this._extensionManager,
+      servicesManager: this._servicesManager,
+      commandsManager: this._commandsManager,
+    };
+
+    previousWorkflowStep?.onBeforeInactivate?.(appContext);
```
Comment (sedghi) [inline platform/core/src/services/WorkflowStepsService/WorkflowStepsService.ts:192]: onBeforeInactivate is a bit odd, don't have suggestions though

File: extensions/cornerstone/src/utils/interleaveTopToBottom.ts [hunk 3]
```diff
@@ -37,6 +37,25 @@ export default function interleaveTopToBottom({
...
+      if (!displaySetOptions?.options?.skipLoading) {
+        numDisplaySetsToLoad++;
+        displaySetsToLoad.add(displaySetInstanceUID);
+      }
+    });
+
+    if (numDisplaySetsToLoad) {
+      filteredMatchDetails.push(curMatchDetails);
+    }
+  });
```
Comment (sedghi) [inline extensions/cornerstone/src/utils/interleaveTopToBottom.ts:57]: can you add a comment on what is happening here?

File: extensions/cornerstone-dynamic-volume/src/getSopClassHandlerModule.ts [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+  const segmentations = segmentationService.getSegmentations();
+  const { seriesMetadata, instance } = _getInstanceFromSegmentations(
+    segmentations
+  );
+
+  // An event is triggered after adding the instance and the displaySet is created
+  DicomMetadataStore.addSeriesMetadata([seriesMetadata], true);
+  DicomMetadataStore.addInstances([instance], true);
+}
```
Comment (sedghi) [inline extensions/cornerstone-dynamic-volume/src/getSopClassHandlerModule.ts:322]: I'm surprised this exists here, why the sopHandler for the line chart should care about segmentation?

File: platform/ui/src/components/LineChart/LineChart.tsx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import classnames from 'classnames';
+import { lineChart } from './d3LineChart';
+import * as d3 from 'd3';
+import './LineChart.css';
+
+const LineChart = ({
+  width: widthProp,
+  height: heightProp,
+  axis,
+  series,
```
Comment (sedghi) [inline platform/ui/src/components/LineChart/LineChart.tsx:12]: I'm surpised there is a prop called series here... is this DICOM series?

File: extensions/cornerstone-dynamic-volume/src/components/LineChartViewport/LineChartViewport.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import { LineChart } from '@ohif/ui';
+
+const LineChartViewport = ({ displaySets }) => {
+  const displaySet = displaySets[0];
+  const { axis: chartAxis, series: chartSeries } = displaySet.chartData;
```
Comment (sedghi) [inline extensions/cornerstone-dynamic-volume/src/components/LineChartViewport/LineChartViewport.tsx:6]: again not sure if series is DICOM series or what

**Review by sedghi — COMMENTED** [review]
see my comments please

**Comment by lscoder** [comment]
*Latest change (38aab5194)*
User is now able to make changes to segmentations and see the chart updated after moving back to Kinetic Analysis.

**Comment by lscoder** [comment]
*Commit 294bcfa*
Changes x-axis unit when more than 50% of the time points are greater than the divisors (1000 ms to s, 60 s  to  m, 60 m to h) making it easier to analyze the data

*Before*
Screenshot_20230831_132921

*After*
Screenshot_20230831_132747

File: platform/ui/src/components/LineChart/LineChart.tsx (hunk 5 above)
Comment (lscoder) [reply platform/ui/src/components/LineChart/LineChart.tsx:12]: No, chart series

File: extensions/cornerstone-dynamic-volume/src/components/LineChartViewport/LineChartViewport.tsx (hunk 6 above)
Comment (lscoder) [reply extensions/cornerstone-dynamic-volume/src/components/LineChartViewport/LineChartViewport.tsx]: `chartData` does not contain any DICOM related data but only label and points.

File: platform/ui/package.json [hunk 7]
```diff
@@ -33,6 +33,13 @@
   "dependencies": {
     "browser-detect": "^0.2.28",
     "classnames": "^2.3.2",
+    "d3-array": "3",
+    "d3-axis": "3",
+    "d3-scale": "4",
+    "d3-scale-chromatic": "3",
+    "d3-selection": "3",
+    "d3-shape": "3",
+    "d3-zoom": "3",
```
Comment (lscoder) [inline platform/ui/package.json:42]: Copied from [D3/package.js](https://github.com/d3/d3/blob/main/package.json)

**Pull request merged**